Entities/Turret.cpp

```diff
@@ -1,395 +1,84 @@
-
-    Attachable::UpdateChildMOIDs(MOIDIndex, m_RootMOID, makeNewMOID);
-}
-
-//////////////////////////////////////////////////////////////////////////////////////////
-// Virtual method:  GetMOIDs
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Puts all MOIDs associated with this MO and all it's descendants into MOIDs vector
-// Arguments:       Vector to store MOIDs
-// Return value:    None.
-
-void Turret::GetMOIDs(std::vector<MOID> &MOIDs) const
-{
-	if (m_pMountedMO && m_pMountedMO->GetsHitByMOs())
-		m_pMountedMO->GetMOIDs(MOIDs);
-
-	Attachable::GetMOIDs(MOIDs);
-}
-
-
-//////////////////////////////////////////////////////////////////////////////////////////
-// Virtual method:  Draw
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Draws this Turret's current graphical representation to a
-//                  BITMAP of choice.
-
-void Turret::Draw(BITMAP *pTargetBitmap,
-               const Vector &targetPos,
-               DrawMode mode,
-               bool onlyPhysical) const
-{
-    if (m_pMountedMO && !m_pMountedMO->IsDrawnAfterParent())
-        m_pMountedMO->Draw(pTargetBitmap, targetPos, mode, onlyPhysical);
-
-    Attachable::Draw(pTargetBitmap, targetPos, mode, onlyPhysical);
-
-    if (m_pMountedMO && m_pMountedMO->IsDrawnAfterParent())
-        m_pMountedMO->Draw(pTargetBitmap, targetPos, mode, onlyPhysical);
-}
-
-} // namespace RTE
\ No newline at end of file
+	ConcreteClassInfo(Turret, Attachable, 20)
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void Turret::Clear() {
+		m_MountedDevice = 0;
+		m_MountedDeviceRotOffset = 0;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	int Turret::Create(const Turret &reference) {
+		if (reference.m_MountedDevice) {
+			m_ReferenceHardcodedAttachableUniqueIDs.insert(reference.m_MountedDevice->GetUniqueID());
+			SetMountedDevice(dynamic_cast<Attachable *>(reference.m_MountedDevice->Clone()));
+		}
+		Attachable::Create(reference);
+
+		m_MountedDeviceRotOffset = reference.m_MountedDeviceRotOffset;
+
+		return 0;
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	int Turret::ReadProperty(std::string propName, Reader &reader) {
+		if (propName == "MountedDevice") {
+			RemoveAttachable(m_MountedDevice);
+			const Entity *mountedDeviceEntity = g_PresetMan.GetEntityPreset(reader);
+			if (mountedDeviceEntity) {
+				m_MountedDevice = dynamic_cast<HeldDevice *>(mountedDeviceEntity->Clone());
+				AddAttachable(m_MountedDevice);
+				m_MountedDevice->SetInheritsRotAngle(false);
+				m_MountedDevice->SetDeleteWhenRemovedFromParent(true);
```

Comment:
Hmmmmm yeah. I'd thought this was prior behaviour, but I can't find anything now that suggests that and I think I was wrong. I guess it should be added as a particle instead of an item so it can't be picked up by actors?

Reply:
Does adding it as a particle actually make it not pickable by actors?
Might need to recreate it as a MOSR/Gib when adding it to MovableMan or something.

Reply:
Looks like it does, in a rather silly way. AHuman determines its ItemInReach by casting an MORay and then getting the object from it. The important part is that it then checks if MovableMan::IsDevice(item), which only checks the Items collection. If it's not an item, it gets ignored.

Reply:
That's pretty silly yeah but I guess it makes sense.
What about the HUD though, is it still visible?

Reply:
Stuff in Particles doesn't get a HUD I think, and the ItemInReach thing determines pie slices and all that.

Reply:
Nah, no need.
While it can be cool to be able to pick up a crab gun after destroying the crab there's no correct human offsets defined and unless you automagic it somehow it's gonna end up looking like ass.
Maybe a consideration for the far future:tm:.

Reply:
Okay, so you're good with it normally just deleting when removed?

Reply:
Nope, definitely needs to be thrown out as a gib.
It'd be enough to throw it out as an unpickable device but i'm pretty sure it will not settle unless the max device threshold is exceeded.
Will casting it back to MOSR before adding it to MovableMan magically yeet all it's device properties and let it settle like any other piece of junk?

Reply:
Yeah, fair enough.
I think the best way to do that is go with the flag that tells MovableMan "you should always add this as a particle not an item" and make it always turned on for Turret guns. Also lets people make hardcoded weapons easily not-pickup-able, which is a nice bonus I guess.